### ISSUE TITLE
Forall the things we've tried, still have a problem!

### ISSUE THREAD

**Issue by buddha314** [buddha314/numsuch#71]
This gives the wrong number density, I've isolated it to this one `forall` loop

```chapel
  cursor2.query(r, (fromField, toField, edgeTable, fromField, toField));
  var dom1: domain(1) = {1..0},
      dom2: domain(1) = {1..0},
      indices: [dom1] (int, int),
      values: [dom2] real;

  // This guy is causing problems.  Exterminiate with extreme prejudice
  //forall row in cursor2 {
  for row in cursor2 {
    indices.push_back((
       rows.get(row[fromField])
      ,cols.get(row[toField])
      ));

    if wField == "NONE" {
      values.push_back(1);
    } else {
      values.push_back(row[wField]: real);
    }
  }
```

I think you said you have a solution for it, not sure.

**Comment by Tshimanga** [comment]
Yeah that's the same bit that was giving me problems. I don't have a performant solution for it yet

**Comment by ben-albrecht** [comment]
Are the iterations of this for loop independent? Does the order of `values` or `indices` matter? That will not be maintained with the parallel version (`forall`).

**Comment by Tshimanga** [comment]
@ben-albrecht yes, the iterations are independent but iterations that collide when trying to carry out their `.push_back()`s are colliding destructively and so there is a less than 1 retention most of the time.

**Comment by Tshimanga** [comment]
@ben-albrecht also worth noting that the issue only arises when running on multiple cores (necessary) as Chapel automatically tries to distribute the work but then the collisions make recombination nonconservative. on a single thread it works fine albeit with performance comparable to a vanilla `for` loop.

**Comment by ben-albrecht** [comment]
> on a single thread it works fine albeit with performance comparable to a vanilla for loop.

Note this is always true. When number of available tasks is 1, a `forall` is the same as a `for` loop (with some minimal overhead).

**Comment by Tshimanga** [comment]
Indeed, that last bit was included just for completeness

**Comment by buddha314** [comment]
does `indices` have to have a dmapped() clause?

**Comment by ben-albrecht** [comment]
Here's an idea:

You could take 2 passes over `cursor2`, where the first pass appends to an associative domain and the second pass sets the values within the associative array:

```chpl
// pass 1
var indices: domain(2*int);
forall row in cursor2 with (+ reduce indices) {
  indices += ij;
}

// pass 2
var values: [indices] real;
forall row in cursor2 {
  var value = if wField == "NONE" then 1 else row[wField]: real;
  var ij = (rows.get(row[fromField]), cols.get(row[toField]));
  values[ij] = value;
}

// Unpack associative array into indices and values arrays here
```

You could also try using the parallel-safe [`DistributedBag`](https://chapel-lang.org/docs/master/modules/packages/DistributedBag.html) to get away with doing 1 pass. This would require pushing 3-tuples, `(i, j, value)`, and you'd still need to unpack them into your desired `indices` and `values` array at the end.

**Comment by Tshimanga** [comment]
@buddha314 I'll try and test this when I get home if you haven't gotten to it already

**Comment by ben-albrecht** [comment]
@Tshimanga - you posted the following code/timings in Gitter:

```chpl
use Time;


var indices1: domain(2*int);
var indices2: domain(2*int);

var t1: Timer;
var t2: Timer;

t1.start();
for i in 1..10000 {
  for j in 1..10000 {
    indices1 += (i,j);
  }
}
t1.stop();


t2.start();
forall i in 1..10000 with (+ reduce indices2) {
  forall j in 1..10000 with (+ reduce indices2) {
    indices2 += (i,j);
  }
}
t2.stop();



writeln("Serial Time: %n".format(t1.elapsed()));
writeln(indices1.size); // ~1.5s 
writeln("Parallel Time: %n".format(t2.elapsed()));
writeln(indices2.size); // ~50s
```

I have a few comments:

1. The nested `forall` statements is not ideal for performance. Each `forall` has a small overhead for checking the available cores for spawning tasks, and spawns at least 1 task to execute its iterations. The available cores will already be saturated with tasks from the outer `forall` loop, so the inner `forall` is creating a lot of unnecessary overhead. It would be better to write the loop with a 2D domain:

```chpl
forall i in {1..10000, 1..10000} with (+ reduce indices2) {
    indices2 += (i,j);
}
```

I suggest you read through the [data parallelism user guide](https://chapel-lang.org/docs/latest/users-guide/datapar/forall.html) for more details.

2. Note that this benchmark is only adding to indices, where the original example is reading from the database as well, which I assume is the most expensive operation. As the cost/iteration increases, the performance gained from parallelizing will increase.

3. How many cores are available on the machine you're testing on?

With a smaller input, here are my timings with 4 cores available:

```chpl
use Time;

config const n = 1000;

var indices1: domain(2*int);
var indices2: domain(2*int);

var t1: Timer;
var t2: Timer;

var D = {1..n, 1..n};
t1.start();
for (i,j) in D {
  indices1 += (i,j);
}
t1.stop();


t2.start();
forall (i,j) in D with (+ reduce indices2) {
  indices2 += (i,j);
}
t2.stop();



writeln("Serial Time: %n".format(t1.elapsed()));
writeln(indices1.size); // 0.359068
writeln("Parallel Time: %n".format(t2.elapsed()));
writeln(indices2.size); // 1.35671
```

**Comment by buddha314** [comment]
Did you want this

```chapel
forall i in {1..10000, 1..10000} with (+ reduce indices2) {
    indices2 += (i,j);
}
```

to read?
```chapel
forall (i,j) in {1..10000, 1..10000} with (+ reduce indices2) {
    indices2 += (i,j);
}
```

**Comment by Tshimanga** [comment]
@ben-albrecht ah yeah not sure why I didn't just use a 2d domain. weird. re-ran the test at n=10,000 and definitely an improvement but still a little over 4 times slower than the `for` loop. Hopefully, however, I can run it on n=100,000 or larger and start to see a performance benefit.

The process is running on 8 cores. Thanks again for the help.

**Comment by Tshimanga** [comment]
Also, earlier I had tried using `ref` instead of `reduce` and with the nested `forall`s that ran in almost a third of the time of `reduce` with a 2d domain. Just ran it again with a 2d domain and it's performance improved further but only slightly. What are your thoughts on `with (ref indices)` @ben-albrecht ?

**Comment by ben-albrecht** [comment]
> What are your thoughts on `with (ref indices)` @ben-albrecht ?

For some reason, I thought this was not parallel-safe, but talking to @mppf makes me reconsider. I think this would be better to use actually. The `+ reduce` would be useful for the associative array loop it there were potential duplicate entries (I have been assuming this is not the case).

**Comment by Tshimanga** [comment]
Yeah `ref` performs roughly thrice as well, but at n=10,000 the loop is still almost twice as slow as serial. That is already pretty large as it's running over 100,000,000 iterations. At n=100,000 the process dies midway. Any other thoughts on getting the parallel version to beat the serial process? Do you think this is worth opening a ticket in chapel-lang/chapel? @ben-albrecht
